csharp/autobuilder/Semmle.Autobuild/DotNetRule.cs

```diff
@@ -258,14 +257,56 @@ CommandBuilder GetRestoreCommand(IBuildActions actions, (string DotNetPath, IDic
             return restore;
         }
 
-        CommandBuilder GetBuildCommand(Autobuilder builder, (string DotNetPath, IDictionary<string, string> Environment)? arg)
+        static BuildScript GetInstalledRuntimesScript(IBuildActions actions, (string DotNetPath, IDictionary<string, string> Environment)? arg)
+        {
+            var listSdks = new CommandBuilder(actions, environment: arg?.Environment, silent: true).
+                RunCommand(DotNetCommand(actions, arg?.DotNetPath)).
+                Argument("--list-runtimes");
+            return listSdks.Script;
+        }
+
+        /// <summary>
+        /// Gets the `dotnet build` script.
+        ///
+        /// The CLR tracer only works on .NET Core >= 3 on Linux and macOS (see
+        /// https://github.com/dotnet/coreclr/issues/19622), so in case we are
+        /// running on an older .NET Core, we disable shared compilation (and
+        /// hence the need for CLR tracing), by adding a
+        /// `/p:UseSharedCompilation=false` argument.
+        /// </summary>
+        BuildScript GetBuildScript(Autobuilder builder, (string DotNetPath, IDictionary<string, string> Environment)? arg, string projOrSln)
         {
             var build = new CommandBuilder(builder.Actions, null, arg?.Environment);
-            return builder.MaybeIndex(build, DotNetCommand(builder.Actions, arg?.DotNetPath)).
+            var script = builder.MaybeIndex(build, DotNetCommand(builder.Actions, arg?.DotNetPath)).
                 Argument("build").
-                Argument("--no-incremental").
-                Argument("/p:UseSharedCompilation=false").
-                Argument(builder.Options.DotNetArguments);
+                Argument("--no-incremental");
+
+            if (builder.Actions.IsWindows())
+                return script.Argument(builder.Options.DotNetArguments).
+                    QuoteArgument(projOrSln).
+                    Script;
+
+            return BuildScript.Bind(GetInstalledRuntimesScript(builder.Actions, arg), (runtimes, runtimesRet) =>
+            {
+                var compatibleClr = false;
+                if (runtimesRet == 0)
+                {
+                    var regex = new Regex(@"Microsoft\.NETCore\.App (\d)");
+                    compatibleClr = runtimes.
+                        Select(runtime => regex.Match(runtime)).
+                        Where(m => m.Success).
+                        Any(m => int.TryParse(m.Groups[1].Value, out var version) && version >= 3);
+                }
+
+                return compatibleClr ?
+                    script.Argument(builder.Options.DotNetArguments).
+                        QuoteArgument(projOrSln).
+                        Script :
+                    script.Argument("/p:UseSharedCompilation=false").
```

Comment:
Do you think it is preferable to use `/p:UseSharedCompilation=false` or to use the environment variable `UseSharedCompilation=false` ? If we have the environment variable set anyway, then do we really need to add this as well?